File: countdown.js
const fs = require('fs');
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const wholeDictionary = require("./english.json");

const dictionary = Object.keys(wholeDictionary);

let alphabet = [...Array(26).keys()].map((i) => String.fromCharCode(i + 97));

const getRandom = (array = alphabet) =>
  array[Math.floor(Math.random() * array.length)];

let letterAmount = 2;

let levelNumber = 0;
//how many words you got right so far

let score = 0;
//your current total score

const possibilities = { words: [], maxLength: 0 };

function countdown() {
  let currentLetters = [];
  for (let i = 0; i < letterAmount; i++) {
    let lettersLeft = alphabet.filter((el) => !currentLetters.includes(el));
    currentLetters.push(getRandom(lettersLeft));
  }

  let lettersStringified = currentLetters.join(", ");

  //this checks if it is actually possible to answer the question.

  possibilities.words = [];
  //this resets the words array

  possibilities.maxLength = 0;
  //this resets the max posible length of a word that fits

  dictionary.forEach((word) => {
    for (letter of currentLetters) {
      if (!word.toLowerCase().split("").includes(letter)) return;
    }
    if (word.length > possibilities.maxLength)
      possibilities.maxLength = word.length;
    return possibilities.words.push(word);
  });
  //checks all possible words that fit

  if (!possibilities.words.length) countdown();
  //if it's impossible to solve, then get a new pair of letters

  rl.question(
    `Please input a word that includes all the following letters. The longer the word, the more points you earn!
    ${lettersStringified}
    
    `,
    function (currentWord) {
      currentWord = currentWord.toLowerCase();
      if (!possibilities.words.includes(currentWord)) {
        rl.close();
      }

      for (letter of currentLetters) {
        if (!currentWord.split("").includes(letter)) rl.close();
      }

      //get to next level

      console.log(
        `Good job! Another word you could have gone with is ${getRandom(
          possibilities.words.filter(
            (w) => w !== currentWord && w.length === possibilities.maxLength
          )
        )}`
      );
      letterAmount = 2 + Math.floor(Math.cbrt(levelNumber));
      levelNumber++;
      score += Math.floor(currentWord.length ** letterAmount / 2);
      countdown();
    }
  );
}

countdown();

rl.on("close", function () {

  const highScore = fs.readFileSync('./highscore.txt', 
  {encoding:'utf8', flag:'r'}); 

  console.log(
    `You could have gone with "${getRandom(
      possibilities.words.filter((w) => w.length === possibilities.maxLength)
    )}" :(... but unfortunately you lost, you got ${levelNumber} words right and your final score is ${score}`
  );


  if (score > highScore)  {
    console.log(`Congratulations on setting a new high score, though! You're now at ${highScore} points`)

    fs.writeFileSync('./highscore.txt', `${score}`);
    //overwrites previous high score
    //maybe one could keep previous high scores in order to make comparisons?

    console.log(`You did set a new highscore though... ${highScore}`)
  }
  else {
    console.log(`Your highscore is still ${highScore}`)
  }

  process.exit(0);
});

File: countdown.py
import json
import string
import random
import math
import os.path
from os import path

with open('english.json') as f:
  whole_dictionary = json.load(f)

dictionary = whole_dictionary.keys()

alphabet = set(string.ascii_lowercase)

letter_amount = 2

level_number = 0

#how many words you got right so far

score = 0
#your current total score

lost = False

possibilities = { "words": [], "maxLength": 0 }

def lose(current_word):
  global lost
  global possibilities
  global level_number
  global score

  lost = True

  if (not path.exists("highscore.txt")):
    with open('highscore.txt', 'w') as f:
      f.write(str(0))
      f.close()
  
  with open('highscore.txt') as f:
      high_score = int(f.read())
      f.close()

  long_possibility = random.choice(list(filter(lambda w: w != current_word and len(w) == possibilities["maxLength"], possibilities["words"])))

  msg = f'You could have gone with {long_possibility} :(... but unfortunately you lost, you got {level_number} words right and your final score is {score}'

  if (score > high_score):
    msg += "... which is now your new high score!"
    with open('highscore.txt', 'w') as f:
      f.write(str(score))
      f.close()
  
  print(msg)

def countdown():
  
  global level_number
  global letter_amount
  global score

  current_letters = []

  for i in range(letter_amount):
    letters_left = alphabet.difference(current_letters)
    current_letters.append(random.choice(list(letters_left)))

  letters_stringified = ' '.join(current_letters)

  #this checks if it is actually possible to answer the question.

  possibilities["words"] = []
  #this resets the words array

  possibilities["maxLength"] = 0
  #this resets the max posible length of a word that fits

  for word in dictionary:
    if (set(word).issuperset(current_letters)):
        if (len(word) > possibilities["maxLength"]):
          possibilities["maxLength"] = len(word)
        possibilities["words"].append(word)
  
  #checks all possible words that fit

  if (not possibilities["words"]):
    print("WAS IMPOSSIBLE TO SOLVE")
    countdown()
  #it's impossible to solve, then get a new pair of letters

  while(not lost):
    current_word = input(f'Please input a word that includes all the following letters. The longer the word, the more points you earn! {letters_stringified} \n').lower()

    if (current_word not in possibilities["words"]):
      return lose(current_word)
    for letter in current_letters:
      if (letter not in current_word):
        return lose(current_word)
        #get to next level

      if(lost):
         return

      another_choice = random.choice(list(filter(lambda w: w != current_word and len(w) == possibilities["maxLength"], possibilities["words"])))

      print(f'Good job! Another word you could have gone with is {another_choice}')

      letter_amount = 2 + math.floor(level_number**(1./3.))
      level_number += 1

      score += math.floor(len(current_word) ** letter_amount / 2)
      countdown()

countdown()